File: code/bot.js
const express = require('express');
const fs = require('fs').promises;
const path = require('path');
const opn = require('opn');
const puppeteer = require('puppeteer');
const { parse } = require('json2csv');

const app = express();
const port = 3000;

app.use(express.json());
app.use(express.static(__dirname));

class Bot {
    constructor() {
        this.browser = null;
        this.page = null;
    }

    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    getRandomDelay(min, max) {
        return Math.floor(Math.random() * (max - min + 1) + min);
    }

    async autoScroll() {
        await this.page.evaluate(async () => {
            await new Promise((resolve) => {
                var totalHeight = 0;
                var distance = 100;
                var timer = setInterval(() => {
                    var scrollHeight = document.body.scrollHeight;
                    window.scrollBy(0, distance);
                    totalHeight += distance;

                    if (totalHeight >= scrollHeight - window.innerHeight) {
                        clearInterval(timer);
                        resolve();
                    }
                }, 100);
            });
        });
    }

    async login(username, password) {
        try {
            await this.page.goto("https://www.facebook.com/", { waitUntil: 'networkidle2' });
            await this.page.type("[aria-label='Email or phone number']", username, { delay: 30 });
            await this.page.type("[aria-label='Password']", password, { delay: 30 });
            await this.page.click("[name='login']");
            await this.page.waitForNavigation({ waitUntil: 'networkidle2' });
        } catch (error) {
            console.error('Error logging in:', error);
        }
    }

    async scrapeGroupProfileURLs(group_url) {
        await this.page.goto(group_url, { waitUntil: 'networkidle2' });
        await this.page.waitForSelector("[class='x1i10hfl xjbqb8w x1ejq31n xd10rxx x1sy0etr x17r0tee x972fbf xcfux6l x1qhh985 xm0m39n x9f619 x1ypdohk xt0psk2 xe8uvvx xdj266r x11i5rnm xat24cr x1mh8g0r xexx8yu x4uap5 x18d9i69 xkhd6sd x16tdsg8 x1hl2dhg xggy1nq x1a2a7pz x1sur9pj xkrqix3 xzsf02u x1s688f']", { timeout: 10000 });
        // await this.autoScroll();
        // await this.sleep(2000);
        // await this.autoScroll();
        // await this.sleep(2000);
        // await this.autoScroll();

        const groupProfileURLs = await this.page.evaluate((selector) => {
            const elements = document.querySelectorAll(selector);
            const urls = new Set();
            elements.forEach(element => {
                const href = element.href;
                if (href && href.includes('groups')) {
                    urls.add(href);
                }
            });
            return Array.from(urls);
        }, "[class='x1i10hfl xjbqb8w x1ejq31n xd10rxx x1sy0etr x17r0tee x972fbf xcfux6l x1qhh985 xm0m39n x9f619 x1ypdohk xt0psk2 xe8uvvx xdj266r x11i5rnm xat24cr x1mh8g0r xexx8yu x4uap5 x18d9i69 xkhd6sd x16tdsg8 x1hl2dhg xggy1nq x1a2a7pz x1sur9pj xkrqix3 xzsf02u x1s688f']");
    
        return groupProfileURLs;
    }

    async scrapeProfileData(number_of_hours, group_url) {
        const groupProfileURLs = await this.scrapeGroupProfileURLs(group_url);
        const totalProfiles = groupProfileURLs.length;
        const totalSecondsInDay = number_of_hours * 60 * 60;
        const delayBetweenProfiles = totalSecondsInDay / totalProfiles * 1000;
        console.log("Unique profile URLs to scrape:");
        console.log(groupProfileURLs.length);
        await this.page.waitForSelector("[class='x1i10hfl xjbqb8w x1ejq31n xd10rxx x1sy0etr x17r0tee x972fbf xcfux6l x1qhh985 xm0m39n x9f619 x1ypdohk xt0psk2 xe8uvvx xdj266r x11i5rnm xat24cr x1mh8g0r xexx8yu x4uap5 x18d9i69 xkhd6sd x16tdsg8 x1hl2dhg xggy1nq x1a2a7pz x1sur9pj xkrqix3 xzsf02u x1s688f']", { timeout: 10000 });
        const profileData = [];
    
        for (const groupProfileURL of groupProfileURLs) {
            try {
                console.log(`Waiting for ${delayBetweenProfiles / 1000} seconds before next profile...`);
                await this.sleep(delayBetweenProfiles);
    
                await this.page.goto(groupProfileURL, { waitUntil: 'networkidle2' });
                await this.page.waitForSelector("[aria-label='View profile']", { timeout: 10000 });
    
                const profile_url = await this.page.$eval("[aria-label='View profile']", element => element.href);
                await this.page.goto(profile_url, { waitUntil: 'networkidle2' });
                await this.page.waitForSelector("[class='x193iq5w xeuugli x13faqbe x1vvkbs x1xmvt09 x1lliihq x1s928wv xhkezso x1gmr53x x1cpjm7i x1fgarty x1943h6x x14qwyeo xw06pyt x579bpy xjkpybl x1xlr1w8 xzsf02u x2b8uid']", { timeout: 5000 });
    
                const profile_name = await this.page.$eval("[class='x193iq5w xeuugli x13faqbe x1vvkbs x1xmvt09 x1lliihq x1s928wv xhkezso x1gmr53x x1cpjm7i x1fgarty x1943h6x x14qwyeo xw06pyt x579bpy xjkpybl x1xlr1w8 xzsf02u x2b8uid']", element => element.textContent.trim());

                await this.page.waitForSelector("[class='x193iq5w xeuugli x13faqbe x1vvkbs x1xmvt09 x1lliihq x1s928wv xhkezso x1gmr53x x1cpjm7i x1fgarty x1943h6x xudqn12 x3x7a5m x1f6kntn xvq8zen xo1l8bm xzsf02u x1yc453h']", { timeout: 5000 });
    
                const marital_status = await this.page.evaluate((selector) => {
                    const elements = document.querySelectorAll(selector);
                    for (let element of elements) {
                        const text = element.textContent.trim().toLowerCase();
                        if (text.includes('married')) {
                            return text;
                        }
                    }
                    return 'Not specified';
                }, "[class='x193iq5w xeuugli x13faqbe x1vvkbs x1xmvt09 x1lliihq x1s928wv xhkezso x1gmr53x x1cpjm7i x1fgarty x1943h6x xudqn12 x3x7a5m x1f6kntn xvq8zen xo1l8bm xzsf02u x1yc453h']");
    
                profileData.push({ profile_name: profile_name, profile_url: profile_url, marital_status: marital_status });
            } catch (error) {
                console.error('Error scraping profile: ${groupProfileURL}, error');
            }
        }
    
        await this.saveToCsv(profileData);
    }

    async saveToCsv(data) {
        try {
            const csv = parse(data);
            await fs.writeFile('profile_data.csv', csv);
            console.log('CSV file saved successfully.');
        } catch (error) {
            console.error('Error saving CSV file:', error);
        }
    }

    async runBot(username, password, group_url, number_of_hours) {
        try {
            await new Promise(resolve => setTimeout(resolve, 2000));

            this.browser = await puppeteer.launch({
                headless: false,
                args: [
                    '--no-sandbox',
                    '--disable-setuid-sandbox',
                    '--disable-notifications'
                ]
            });
            this.page = await this.browser.newPage();
            await this.login(username, password);

            await this.scrapeProfileData(number_of_hours, group_url);

        } catch (error) {
            console.error('An error occurred:', error);
        }
        this.browser.close(); //finally {
            // if (this.browser) {
                // await 
            // }
        // }
    }
}

app.get('/', (req, res) => {
    res.sendFile(path.join(__dirname, 'index.html'));
});

app.post('/run-bot', async (req, res) => {
    const { username, password, group_url, number_of_hours } = req.body;

    console.log('Received request to run bot with:', { username, password, group_url, number_of_hours });

    try {
        const bot = new Bot();
        await bot.runBot(username, password, group_url, number_of_hours);
        res.send('Bot execution completed successfully!');
    } catch (error) {
        console.error('Error running bot:', error);
        res.status(500).send('Error running bot');
    }
});

app.listen(port, () => {
    console.log(`Server running at http://localhost:${port}`);
    opn(`http://localhost:${port}`);
});

File: deprecated/bot.py
#%%

from flask import Flask, request, send_from_directory
import os
import time
import random
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager
import json
from threading import Timer
import pandas as pd
import webbrowser

app = Flask(__name__)
port = 3000

class Bot:
    def __init__(self, response):
        self.browser = None
        self.page = None
        self.response = response
        self.profile_data_path = 'profile_data.csv'
        if os.path.exists(self.profile_data_path):
            self.profile_data = pd.read_csv(self.profile_data_path)
        else:
            self.profile_data = pd.DataFrame()
        with open('data.json', 'r') as file:
            self.data = json.load(file)

    def get_random_delay(self, min, max):
        return random.randint(min, max)

    def auto_scroll(self):
        print('scrolling...')
        last_height = self.page.execute_script("return document.body.scrollHeight")
        while True:
            self.page.execute_script("window.scrollBy(0, 100);")
            time.sleep(1)
            
            new_height = self.page.execute_script("return document.body.scrollHeight")
            if new_height == last_height:
                break
            last_height = new_height
        print('done scrolling')

    def login(self):
        self.page.get("https://www.facebook.com/")
        self.page.find_element(By.CSS_SELECTOR, self.data['username_selector']).send_keys(self.response.get('username'))
        self.page.find_element(By.CSS_SELECTOR, self.data['password_selector']).send_keys(self.response.get('password'))
        self.page.find_element(By.CSS_SELECTOR, self.data['login_selector']).click()
        time.sleep(5)

    def scrape_group_profile_urls(self):
        print(f"Opening group URL {self.response.get('group_url')}")
        self.page.get(self.response.get('group_url'))
        time.sleep(10)
        # self.auto_scroll()
        # self.auto_scroll()

        elements = self.page.find_elements(By.CSS_SELECTOR, self.data['group_members_selector'])
        self.group_profile_urls = set()
        for element in elements:
            href = element.get_attribute('href')
            if href and 'groups' in href:# and href not in self.profile_data['group_profile_urls']
                # if len(self.profile_data) > 0:
                    # if not href in self.profile_data['group_profile_urls']:
                self.group_profile_urls.add(href)
        self.group_profile_urls = list(self.group_profile_urls)

    def scrape_profile_data(self):
        print("Unique profile URLs to scrape:", len(self.group_profile_urls))

        total_run_time_in_seconds = float(self.response.get('hours_to_scrape')) * 60 * 60
        time_between_scrapes_in_seconds = total_run_time_in_seconds / float(self.response.get('profiles_to_scrape'))
        print(f'time_between_scrapes_in_seconds: {time_between_scrapes_in_seconds}')

        print(f"Scraping {str(self.response.get('profiles_to_scrape'))} pages over {str(self.response.get('hours_to_scrape'))} hours")

        pages_scraped = 0
        start_time = time.time()

        for group_profile_url in self.group_profile_urls[:int(self.response.get('profiles_to_scrape'))]:
            print(f'getting...{group_profile_url}')
            elapsed_time = time.time() - start_time
            print(f'elapsed_time: {elapsed_time}')
            if elapsed_time > total_run_time_in_seconds:
                print('Total bot runtime exceeded. Ending process.')
                break

            print(f"Waiting for {time_between_scrapes_in_seconds} seconds before next profile...")
            time.sleep(time_between_scrapes_in_seconds)

            self.page.get(group_profile_url)
            time.sleep(10)

            profile_url = self.page.find_element(By.CSS_SELECTOR, self.data['profile_url_selector']).get_attribute('href')
            self.page.get(profile_url)
            time.sleep(5)

            profile_name = self.page.find_element(By.CSS_SELECTOR, self.data['profile_name_selector']).text.strip()
            print(f"Scraping...{profile_name}")

            marital_status = 'not specified'
            elements = self.page.find_elements(By.CSS_SELECTOR, self.data['marital_status_selector'])
            for element in elements:
                text = element.text.strip().lower()
            if 'married' in text:
                marital_status = 'married'
            elif 'single' in text:
                marital_status = 'single'

            # new_row = { 'profile_name': profile_name, 'marital_status': marital_status, 'group_profile_url': group_profile_url, 'profile_url': profile_url }
            # self.profile_data.append(new_row)

            # if marital_status != 'not specified':
            print('marital_status: ', marital_status)
            self.profile_data.loc[len(self.profile_data.index)] = [profile_name, marital_status, group_profile_url, profile_url] 
            self.profile_data.to_csv(self.profile_data_path, index=False)
            print(self.scraped_data)

        pages_scraped += 1

    def open_browser(self):
        print("Starting Selenium...")
        options = Options()
        options.add_argument('--no-sandbox')
        options.add_argument('--disable-setuid-sandbox')
        options.add_argument('--disable-notifications')
        # self.browser = webdriver.Chrome(service=Service(ChromeDriverManager().install()), options=options)
        self.browser = webdriver.Chrome(options=options)
        print("Opening browser...")
        self.page = self.browser
        print("Logging in...")

    def run_bot(self):
        self.open_browser()
        self.login()
        self.scrape_group_profile_urls()
        self.scrape_profile_data()

# @app.route('/')
# def index():
#     return send_from_directory(os.getcwd(), 'index.html')

# @app.route('/run-bot', methods=['POST'])
# def run_bot():
# 	try:
# 		print('running bot...')
# 		bot = Bot(request.json)
# 		bot.run_bot()
# 		return 'Bot execution completed successfully!', 200
# 	except Exception as error:
# 		print('Error running bot:', error)
# 		return 'Error running bot', 500
# 	finally:
# 		close_server()

# def close_server():
#     print('Closing server...')
#     os._exit(0)

# if __name__ == '__main__':
#     # webbrowser.open(url)
#     # Timer(1, open_browser).start()
#     url = f'http://localhost:{port}'
#     webbrowser.open(url)
#     app.run(port=port)

bot = Bot({'username': 'porterbmoody@gmail.com', "password":'Yoho1mes', "group_url":'https://www.facebook.com/groups/344194813025772/members/',
           'profiles_to_scrape':'5', 'hours_to_scrape':'.005'})
bot.run_bot()


#%%
